skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BluetoothListAdapter.java

```diff
@@ -0,0 +1,142 @@
+import android.bluetooth.BluetoothDevice;
+import android.content.Context;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+
+import org.odk.share.R;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+
+import static android.bluetooth.BluetoothDevice.BOND_BONDED;
+
+
+/**
+ * BluetoothListAdapter: {@link RecyclerView.Adapter} for {@link BluetoothDevice}.
+ *
+ * @author huangyz0918 (huangyz0918@gmail.com)
+ * @since 12/06/2019
+ */
+public class BluetoothListAdapter extends RecyclerView.Adapter<BluetoothListAdapter.ViewHolder> {
+
+    private List<BluetoothDevice> bluetoothDeviceList;
+    private OnDeviceClickListener btDeviceClickListener;
+
+    public BluetoothListAdapter(OnDeviceClickListener btDeviceClickListener) {
+        this.btDeviceClickListener = btDeviceClickListener;
+        this.bluetoothDeviceList = new ArrayList<>();
+    }
+
+    /**
+     * Methods for updating data.
+     */
+    public void setDevices(List<BluetoothDevice> dateSet) {
+        this.bluetoothDeviceList = dateSet;
+    }
+
+    public void addDevice(BluetoothDevice bluetoothDevice) {
+        if (!bluetoothDeviceList.contains(bluetoothDevice)) {
+            this.bluetoothDeviceList.add(bluetoothDevice);
+        }
+        notifyDataSetChanged();
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull final ViewHolder viewHolder, int position) {
+        Context context = viewHolder.itemView.getContext();
+        BluetoothDevice device = bluetoothDeviceList.get(position);
+        String deviceName = device.getName();
+        String deviceAddress = device.getAddress();
+        int deviceBondState = device.getBondState();
+        viewHolder.deviceName.setText(deviceName == null ? viewHolder.itemView.getResources().getString(R.string.bluetooth_instance_name_default) : deviceName);
+        viewHolder.deviceAddress.setText(String.format("%s (%s)", deviceAddress,
+                deviceBondState == 10 ? context.getString(R.string.bluetooth_unpaired) : context.getString(R.string.bluetooth_paired)));
+        if (device.getBondState() == BOND_BONDED) {
+            viewHolder.deviceLogo.setImageResource(R.drawable.ic_smart_phone_yellow);
+        }
+    }
+
+    @Override
+    public int getItemCount() {
+        return bluetoothDeviceList.size();
+    }
+
+    @NonNull
+    @Override
+    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_bt_device, null);
+        return new ViewHolder(view);
+    }
+
+    /**
+     * Rescan the bluetooth devices and update the list.
+     */
+    public void updateDeviceList() {
```

Comment:
Lets not keep this method as well in adapter, try to keep it in some utility class or the activity it is used in.

Reply:
sure